cpp/src/arrow/flight/server.cc

```diff
@@ -791,16 +799,30 @@ struct FlightServerBase::Impl {
 
   void DoHandleSignal(int signum) {
     got_signal_ = signum;
+#ifdef _WIN32
     server_->Shutdown();
+#else
+    int saved_errno = errno;
+    // Ignore errors - pipe is nonblocking
+    write(self_pipe_[1], "0", 1);
+    errno = saved_errno;
+#endif
   }
+
+#ifndef _WIN32
+  static void WaitForSignals(int fd) {
+    // Wait for a signal handler to write to the pipe
+    int8_t buf[1];
+    read(fd, /*buf=*/buf, /*count=*/1);
```

Comment:
This could be fail with `EINTR`. Of course, presumably that's not really an error here :-)